### ISSUE TITLE
SearchGuard performance when fetching large documents

### ISSUE THREAD

**Issue by dorony** [floragunncom/search-guard#310]
Elasticsearch version: 5.2.1
Searchguard version: 5.2.1-11
JVM version: 1.8.0_121
OS version: Ubuntu 14.04.5
Number of nodes: 8

Hi,
We're seeing a big performance difference between a SearchGuard protected cluster and a vanilla ES5 cluster.
The difference comes into play in both search and fetch (multi-get by id), mainly when fetching large documents (~100K each). We tested against two identical clusters, the only difference being the installation of the search-guard plugin. We are seeing up to 50% difference in round trip time between the two clusters, while the reported "took" time by ElasticSearch is the same. We configured SearchGuard to use OpenSSL.

Anything we're missing? Is this the expected performance overhead of encryption?

**Comment by jochenkressin** [comment]
Hi - no, this is not expected behavior. The impact of TLS encryption depends on many factors, but we do have customers with large documents (up to 20K fields per document) and have not heard about such a huge performance impact yet.

What kind of hardware and cipher suites do you use? Does your CPU support hardware encryption, like Intel Xeon or Intel Core? 

Which OpenSSL version do you use?

Also, which features of Search Guard do you use? If you enabled Document- and Field-level security, this can have a performance impact too under certain conditions.

**Comment by dorony** [comment]
1. We are using the default cipher suites, certificates were created according to the instructions [here](https://github.com/floragunncom/search-guard-ssl-docs/blob/master/certificates.md).
1. We are using standard aws machines - m4.2xlarge. They use Xeon processes as far as I know.
1. OpenSSL version - OpenSSL 1.0.1f 6 Jan 2014 (yes a bit old, but this is a development environment only)
1. We're not using document/field level security, only index level.

**Comment by floragunncom** [comment]
Can you elaborate a bit more on how you did your measurements? Can you share a reproducer or at least parts of your perf test code/scripts (if you cannot share them in the public just send us a mail)? If you also use https please disable ssl for http and check if this change anything.

Maybe the overhead is not ssl related but related to authentication and authorization. So it would be great if you can provide your elasticsearch.yml and your sg_config.yml.

**Comment by dorony** [comment]
Below is python code that generates an index with a single large document, and then fetches it 10 times. The result I'm seeing:
```
ES with SG with SSL:
total time = 5334.09404755ms
ES with SG without ssl:
total time = 4786.1161232ms
ES without SG:
total time = 3046.77391052ms
```

```
import time
import json
import elasticsearch

INDEX_NAME = "temp_index"
TYPE = "type1"

es = Elasticsearch(...)

def generate_large_json():
    data = {i: i for i in xrange(0, 300000)}
    return json.dumps({"data": data})


def create_index_with_temp_data():
    large_json = generate_large_json()

    mapping = {
        "settings": {
            "number_of_shards": 1,
            "number_of_replicas": 0
        },
        "mappings": {
            TYPE: {
                "properties": {
                    "data": {"type": "object",
                             "enabled": False}
                }
            }
        }
    }
    es.indices.create(INDEX_NAME, body=mapping)
    es.index(INDEX_NAME, TYPE, large_json, id="data1")

create_index_with_temp_data()
start = time.time()
for i in xrange(10):
    es.get(INDEX_NAME, "data1")
end = time.time()

total_time_ms = (end-start)*1000
print("total time = {}ms".format(total_time_ms))
```

I will send you elasticsearch.yml and and sg_config.yml over email.

**Comment by floragunncom** [comment]
Are these results reproducible when you execute the script multiple times?

Pls. disable OpenSSL for transport and http and share the results.

**Comment by dorony** [comment]
Yes, results are consistent across multiple runs.
I disabled OpenSSL and performance is a little worse - hovering around 5.5 seconds

**Comment by floragunncom** [comment]
Can you enable debug logging and restart elasticsearch and send us the _complete_ logs (from startup till the point where elasticsearch is fully up). There should be something in there like:

<pre>
....
[DEBUG][i.n.u.i.PlatformDependent] Java version: 8
[DEBUG][i.n.u.i.PlatformDependent] -Dio.netty.noJavassist: false
[DEBUG][i.n.u.i.PlatformDependent] Javassist: unavailable
[DEBUG][i.n.u.i.PlatformDependent] You don't have Javassist in your class path or you don't have enough permission to load dynamically generated classes.  Please check the configuration for better performance.
[DEBUG][i.n.u.i.PlatformDependent] -Dio.netty.tmpdir: /var/folders/8s/bwpvj6xs22sd9j0fll3fd0440000gp/T (java.io.tmpdir)
[DEBUG][i.n.u.i.PlatformDependent] -Dio.netty.bitMode: 64 (sun.arch.data.model)
[DEBUG][i.n.u.i.PlatformDependent] -Dio.netty.noPreferDirect: true
[DEBUG][i.n.u.i.PlatformDependent] io.netty.maxDirectMemory: -1 bytes
</pre>

You can enable debug logging in config/log4j2.properties setting `rootLogger.level = debug`

**Comment by floragunncom** [comment]
And make sure your system have enough entropy. Check /proc/sys/kernel/random/entropy_avail and use something like 

* haveged - http://www.issihosts.com/haveged/
* rngd -r /dev/urandom
* ...

if your entropy is too low. See also https://www.cyberciti.biz/open-source/debian-ubuntu-centos-linux-setup-additional-entropy-for-server-using-aveged-rng-tools-utils/

**Comment by dorony** [comment]
Sent you an email with the log.
Entropy is around 900.

**Comment by floragunncom** [comment]
I look at the logs. Meanwhile i suggest you try to raise your entropy level, 900 is not really bad but its also not really good. Try to get it > 2000.

**Comment by floragunncom** [comment]
I see that x-pack is installed alongside with Search Guard. This *may* work but it's not supported. Can you remove x-pack and see if this change anything? (we do not test such setups and therefore its not supported)

**Comment by floragunncom** [comment]
can you post the result of `sudo cat /proc/sys/net/core/somaxconn` ?

**Comment by floragunncom** [comment]
How does your certificates/keys look like? RSA keysize?

If you use RSA based certificates/keys maybe its worthy to try ECDSA.

Generate your key/cert like
<pre>
openssl ecparam -name prime256v1 -genkey -param_enc explicit -out private-key.pem 
openssl req -new -x509 -key private-key.pem -out server.pem -days 730 -sha256
</pre>

and for best performance limit enabled ssl cipher suites to `ECDHE-ECDSA-AES128-GCM-SHA256`

If you have more time i'd recommend to compile openssl with ChaCha20-Poly1305 support and use an ECDSA certificate with ´ECDHE-ECDSA-CHACHA20-POLY1305´ as enabled cipher suite. Fot this you need to patch the 1.0.2 openssl codebase, see https://www.pckr.co.uk/openssl-chacha20-cipher-suite/

**Comment by dorony** [comment]
`cat /proc/sys/net/core/somaxconn` = 128
Regarding certificate/keys, we used your defaults, as suggested [here](https://github.com/floragunncom/search-guard-ssl-docs/blob/master/certificates.md), so RSA key size is 2048. 

I will try your suggestions, but just to understand - did you try running my benchmark against any other cluster and got different results? It would help as it will tell us if the problem is with my configuration or not.

**Comment by dorony** [comment]
I implemented your suggestions: removed x-pack, increased entropy (now above 2000), and moved to ECDSA keys. The only step that had any effect was using ECDSA, which brought me to around 5 seconds in the benchmark. Still much higher than the vanilla ES result (3 seconds)

**Comment by floragunncom** [comment]
After digging more in this it seems like a netty problem. I opened a issue https://github.com/netty/netty/issues/6571

Can you determine the approx. document size from which the performance drop seems to be more than 5-10% compared to non-ssl performance (assume for smaller documents there is no problem).

**Comment by floragunncom** [comment]
Can you run https://github.com/floragunncom/netty-memleak/releases/download/slow1/netty-sslslow-1.0-jar-with-dependencies.jar on one of your machine and post the results (just the line `Took ??? ms`):

`java -Xmx4g -jar netty-sslslow-1.0-jar-with-dependencies.jar`
and
`java -Xmx4g -Dnossl -jar netty-sslslow-1.0-jar-with-dependencies.jar`

**Comment by dorony** [comment]
With SSL: Took 2023 ms
No SSL: Took 769 ms
I do see some errors in your script about loading tcnative (Failed to load netty-tcnative; OpenSslEngine will be unavailable...) - not sure if you intended OpenSSL to work here.

**Comment by floragunncom** [comment]
errors are expected so you can just ignore them. 

Just wonder why your figures are lower than mine (1200 ms/400ms on Macbook with 4 cores I7 2,5 ghz with java 1.8.0_102)

**Comment by dorony** [comment]
Not that surprising, I ran the test on an Amazon VM, I'd expect the single core performance of your mac to be faster

**Comment by floragunncom** [comment]
Netty seems to make progress https://github.com/netty/netty/issues/6571#issuecomment-302495256

**Comment by floragunncom** [comment]
netty 4.1.14 claims to have some improvements ...

**Comment by dorony** [comment]
hi, any progress on this one?

**Comment by floragunncom** [comment]
did you try if recent netty versions improves the situation?

**Comment by dorony** [comment]
It appears the latest ES version still uses 4.1.13, so I guess not.

**Comment by floragunncom** [comment]
Can you pls post the full python testcode you use?

**Comment by dorony** [comment]
I posted it above, it is the full code...

**Comment by floragunncom** [comment]
With netty 4.1.16 (more recent versions are not yet compatible) i see the following figures now

No SSL: 1798 ms
Java SSL: 3081 ms
OpenSSL: 1950 ms

(t2.xlarge, single node with ES 6.1.1 and SG 20.1)

So with OpenSSL there seems to almost no overhead anymore. With Java SSL it appears to be 30% slower.

To replace netty cd into the search-guard plugin folder and

<pre>
rm -rf netty-*4.1*
rm -rf netty-tcnat*
wget http://central.maven.org/maven2/io/netty/netty-handler/4.1.16.Final/netty-handler-4.1.16.Final.jar
wget http://central.maven.org/maven2/io/netty/netty-buffer/4.1.16.Final/netty-buffer-4.1.16.Final.jar
wget http://central.maven.org/maven2/io/netty/netty-codec/4.1.16.Final/netty-codec-4.1.16.Final.jar
wget http://central.maven.org/maven2/io/netty/netty-codec-http/4.1.16.Final/netty-codec-http-4.1.16.Final.jar
wget http://central.maven.org/maven2/io/netty/netty-common/4.1.16.Final/netty-common-4.1.16.Final.jar
wget http://central.maven.org/maven2/io/netty/netty-resolver/4.1.16.Final/netty-resolver-4.1.16.Final.jar
wget http://central.maven.org/maven2/io/netty/netty-transport/4.1.16.Final/netty-transport-4.1.16.Final.jar
</pre>

Then download one of

<pre>
netty-tcnative-openssl-1.0.2n-static-2.0.7.Final-alpine-linux-x86_64.jar
netty-tcnative-openssl-1.0.2n-static-2.0.7.Final-fedora-linux-x86_64.jar
netty-tcnative-openssl-1.0.2n-static-2.0.7.Final-non-fedora-linux-x86_64.jar
</pre> from [https://dl.bintray.com/floragunncom/netty-tcnative](https://dl.bintray.com/floragunncom/netty-tcnative)

and put it into the search-guard plugin folder

I modified the python script to work with python3, it looks now like

```python
import time
import json
import ssl
from elasticsearch import Elasticsearch

#Needs Python3 and elasticsearch

INDEX_NAME = "temp_index"
TYPE = "type1"

es = Elasticsearch(
    ['admin:admin@localhost:9200'],
    #connection_class=RequestsHttpConnection,
    use_ssl=True,
    verify_certs=False,
    #ca_certs='/usr/share/elasticsearch/config/chain-ca.pem',
    ssl_version=ssl.PROTOCOL_TLSv1_2,
    ssl_assert_hostname=False
    #client_key='/usr/share/elasticsearch/config/CN=picard,OU=client,O=client,L=Test,C=DE.key.pem'
    #client_cert='/usr/share/elasticsearch/config/CN=picard,OU=client,O=client,L=Test,C=DE.crtfull.pem',
    )

def generate_large_json():
    data = {i: i for i in range(0, 300000)}
    return json.dumps({"data": data})


def create_index_with_temp_data():
    large_json = generate_large_json()


    mapping = {
        "settings": {
            "number_of_shards": 1,
            "number_of_replicas": 0
        },
        "mappings": {
            TYPE: {
                "properties": {
                    "data": {"type": "object",
                             "enabled": False}
                }
            }
        }
    }
    try:
        es.indices.delete(INDEX_NAME)
    except:
        print("No index to delete")
    es.indices.create(INDEX_NAME, body=mapping)
    es.index(INDEX_NAME, TYPE, large_json, id="data1")

create_index_with_temp_data()
start = time.time()
for i in range(10):
    es.get(INDEX_NAME, TYPE, "data1")
end = time.time()

total_time_ms = (end-start)*1000
print("total time = {}ms".format(total_time_ms))
```

**Comment by floragunncom** [comment]
see also https://github.com/elastic/elasticsearch/pull/28345 and https://github.com/netty/netty/issues/7614

**Comment by floragunncom** [comment]
We did some further ssl perf testing with bulk sizes of 9mb:
<pre>
For SG 6.2.2 with tcnative 2.0.7 openssl static build

- Use a recent version of Java (>= Update 141)
- Prefer ECDSA over RSA keys
- Prefer ECDHE over RSA and ECDH for key exchange
- Make sure you have enough entropie on your system (/proc/sys/kernel/random/entropy_avail > 2000)
- If you have to use Java SSL use one of those ciphers 
  - TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA or
  - TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384
  This comes with a penalty of < 5% compared to non-ssl
  Note: GCM is 6-8 times slower than CBC as Java lacks GCM optimizations
  
- Best is to use OpenSSL with 
  - TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 or 
  - TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 
  This comes with a penalty of < 5% compared to non-ssl

- You should prefer GCM over CBC because its more secure for various reasons.
  Modern browsers also do not longer support CBC mode
  
- Conclusion: If you choose the right crypto parameters and hardware SSL slows down Elasticsearch by approx 3-5%.
  Do not use GCM mode with Java SSL because this dramatically slower. Avoid CBC because browsers do not support it any longer.
</pre>

So i will close this issue for because it seems resolved with netty 4.1.16.Final
